Sources/NIO/BaseSocketChannel.swift

```diff
@@ -917,7 +917,10 @@ class BaseSocketChannel<T: BaseSocket>: SelectableChannel, ChannelCore {
                 becomeActive0(promise: promise)
             }
         } catch let error {
-            promise?.fail(error: error)
+            assert(self.isOpen)
```

Comment:
in other places we opted to go for `assert(self.lifecycleManager.isRegistered)`. Maybe you also want to `assert(!self.lifecycleManager.isActive)` because we should be registered but not active here.

Reply:
That assertion is unnecessary because the `do` block covers it. I suppose I could assert `isRegistered && !isActive`, if you prefer.